### PR TITLE
Fix 0% tax rates not added to refund orders

### LINKED ISSUE #27118: 0% tax rates/amounts are not added to refund orders
**Prerequisites:**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
When refunding an order that has items with a 0% tax rate, taxes are not copied over to the refund order.

**Expected behavior**
0% is not the same as "no" tax, and 0% rates should be copied over to the refund order just like other tax rates.

**Actual behavior**
no 0% rate tax lines are present in the refund order data (`$refund->get_items('tax');`)
no-zero-tax-refund

**Steps to reproduce the bug**
Steps to reproduce the bug:
1. Create an order with a product with 0% tax (e.g. 0.00 showing on the line item)
2. Refund the order (on a line-item level, not just setting to refunded)
3. Check the refund order data - tax for the 0% item is not recorded

**Isolating the problem:**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
WC 4.3.1
WP 5.4.3

### PR DESCRIPTION
**Pull request by mordeth**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

A 0% tax rate is a valid rate (zero-rated VAT, tax-exempt classes), not "no tax", and must be recorded on refunds for accurate tax reporting. When refunding a line item taxed at 0%, the 0% tax line was silently dropped: `$refund->get_items( 'tax' )` held no entry for that rate.

The admin refund UI posts each tax total via `accounting.unformat()` as the numeric `0`, and `wc_format_decimal( 0 )` returns the falsy string `'0'`. Two cooperating callback-less `array_filter()` calls then stripped it — `WC_AJAX::refund_line_items()` before handing off, and `wc_create_refund()` again on its own input — so `update_taxes()` created no `WC_Order_Item_Tax` for the rate.

This keeps every numeric refund-tax entry (including `0`) at both sites. In `wc_create_refund()` the retained values are cast to `float` so the existing `array_map( 'wc_format_refund_total', ... )` calls (whose callback is typed `float`) keep a stable parameter type.

Closes #27118

### How to test the changes in this Pull Request:

1. Enable taxes: **WooCommerce → Settings → General → Enable tax rates and calculations**.
2. Under **WooCommerce → Settings → Tax → Standard rates**, add a rate with **Rate %** `0.0000` and a name like `Zero Rate`.
3. Create a taxable product (e.g. `$20`) in the standard tax class.
4. Place an order containing that product — the order shows a `$0.00` tax line.
5. On the order edit screen, click **Refund**, refund the single line item in full, then **Refund manually**.
6. Inspect the refund's tax items (WP-CLI / snippet):
   ```php
   $refund = current( wc_get_orders( [ 'parent' => <ORDER_ID>, 'type' => 'shop_order_refund' ] ) );
   var_dump( $refund->get_items( 'tax' ) );
   ```
   - **Before:** `array(0) {}` — the 0% tax line is missing.
   - **After:** one `WC_Order_Item_Tax` for `Zero Rate` with `tax_total` `0`.
7. Repeat with a non-zero rate to confirm normal tax refunds are unaffected.

### Testing that has already taken place:

- Reproduced on current `trunk` in wp-env through the faithful admin AJAX refund transform + real `wc_create_refund()`: the refund had **0** tax items. With the fix it has **1** (the 0% rate preserved, `tax_total` `0`).
- Added PHPUnit regression test `WC_Tests_Order_Functions::test_wc_create_refund_preserves_zero_rate_tax_27118`. It **fails on trunk** (0 vs 1 tax item) and **passes with the fix**. The full `WC_Tests_Order_Functions` class is green (46 tests, 1 pre-existing unrelated skip).
- `phpcs` clean on all changed lines; full-project PHPStan clean.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

Created manually — see `plugins/woocommerce/changelog/27118-fix-zero-rate-tax-refund` (patch / fix).

</details>

### Release Communication

Select if this PR needs a generated summary for release notes:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

### CODE REVIEW

File: plugins/woocommerce/includes/wc-order-functions.php [hunk 1]
```diff
@@ -606,7 +606,16 @@ function wc_create_refund( $args = array() ) {
...
+
+				// Keep every numeric tax entry, including a 0% amount that a callback-less array_filter would drop as falsy. 0% is a valid rate, not "no tax". See #27118.
+				$refund_tax = array();
+				if ( isset( $args['line_items'][ $item_id ]['refund_tax'] ) ) {
+					foreach ( (array) $args['line_items'][ $item_id ]['refund_tax'] as $tax_rate_id => $tax_amount ) {
+						if ( is_numeric( $tax_amount ) ) {
+							$refund_tax[ $tax_rate_id ] = (float) $tax_amount;
+						}
+					}
+				}
```
[the remark this answers is not in the document]
Comment (mordeth) [reply plugins/woocommerce/includes/wc-order-functions.php]: Thanks for the careful look here — I checked this against the base commit, and the premise doesn't hold up: `wc_format_decimal()` was never called in `wc_create_refund()`. On base, the line this replaces was a bare callback-less `array_filter()` with no formatting at all (that falsy filter is exactly the bug this PR fixes). The only `wc_format_decimal()` in this flow lives in `WC_AJAX::refund_line_items()`, which this PR preserves — and `git log -S 'wc_format_decimal'` confirms it was never applied to `refund_tax` anywhere in this file's history.

On the backward-compatibility concern — localized inputs like `"1,50"` were never "formatted and accepted" here. Verified empirically on the base commit:

- `"1,50"` passed the old truthiness filter raw and reached `wc_format_refund_total()` (`$amount * -1`), where PHP truncated it to `-1` — a silently corrupted refund tax of 1.00 instead of 1.50.
- Fully non-numeric strings (e.g. `"abc"`) caused a fatal `TypeError` on PHP 8.

Dropping non-numeric entries is strictly safer than either prior outcome, and it matches the validation the REST API path already applies to `refund_tax` (`RestApiParameterUtil` gates it with `is_numeric()`).

One genuine nuance worth acknowledging: on PHP 7.4 (WC's minimum), a trailing-whitespace numeric string like `"1.50 "` fails `is_numeric()` and is now dropped, where base coerced it to `-1.5` with a notice — accepted by accident, not by formatting. On PHP 8.0+, `is_numeric()` accepts surrounding whitespace, so that gap only exists on 7.4 and only for direct callers passing unsanitized strings to a numeric API parameter.

Adding `wc_format_decimal()` here would introduce new store-locale-dependent parsing to `wc_create_refund()`'s public contract rather than restore prior behavior, so I'm keeping the minimal fix as is.

**Review by vladolaru — COMMENTED** [review]
Hey Cvetan — first off, really nice detective work here. Tracing a decade-old bug down to a falsy-`0` `array_filter` is exactly right, the inline comments explaining *why* are genuinely helpful, and `is_numeric` is the right instinct.

I stress-tested the AJAX path across a full **multi-item** order, though, and I think there's a regression we should close before merge. Sharing the whole chain so you can sanity-check my reasoning.

## The issue: untouched items become phantom refund line items

The admin refund JS serializes its three input groups differently (`client/legacy/js/admin/meta-boxes-order.js:943-973`):

- `line_item_qtys` — **gated** by `if ( item.value )` (line 945): untouched qty fields are omitted.
- `line_item_totals` — **ungated** (line 951): every row posts `accounting.unformat( '' ) === 0`.
- `line_item_tax_totals` — **ungated** (line 960): every tax field posts `0` too.

So when a merchant refunds by just typing an amount, or partially refunds one item, on a tax-enabled order, the request already carries `0` totals and `0` taxes for *every* item in the order.

The guard in `wc_create_refund()` (`includes/wc-order-functions.php:620`) is:

```php
if ( empty( $qty ) && empty( $refund_total ) && empty( $args['line_items'][ $item_id ]['refund_tax'] ) ) {
    continue;
}
```

For an untouched item `$qty` is `0` and `$refund_total` is `'0'` (and `empty( '0' )` is `true` in PHP), so the **`refund_tax` leg is the only thing skipping the item.** On trunk the callback-less `array_filter` empties it → item skipped. With `is_numeric`, `refund_tax` becomes `[ rate_id => 0 ]` → non-empty → the guard no longer fires, and **a negative refund line item is created for every item in the order.**

Because a single-item refund has no untouched siblings, this doesn't surface in the one-item repro — it needs a multi-item or amount-only refund.

## Why it matters beyond noise

`wc_create_refund()` adds every surviving line item to `$refunded_order_and_products` with no qty/amount condition (`wc-order-functions.php:627-632`) and then **deletes all customer downloads** for those products (`694-704`). So a partial refund on an order containing downloadable products would **revoke download access for products that were never refunded.** There's also knock-on impact on full-refund analytics attribution and on any REST/3rd-party consumer that divides by a now-`0` refund quantity.

## Suggested fix — server guard (the load-bearing one)

Don't let an all-zero `refund_tax` resurrect an item that has no qty and no total to refund — `wc-order-functions.php:620`:

```php
// An item with no qty and no total shouldn't be resurrected purely by a refund_tax
// array of zeros (e.g. untouched tax fields posted as 0). A genuine 0% tax line still
// rides along because that item is refunded via its qty / refund_total, which keep the
// guard open; only an all-zero, nothing-else-to-refund row is skipped here.
$refund_tax_sum = array_sum( array_map( 'abs', $refund_tax ) );
if ( empty( $qty ) && empty( $refund_total ) && empty( $refund_tax_sum ) ) {
    continue;
}
```

(`$refund_tax` is the already-filtered/cast array built just above at lines 611-618.) This is safe for #27118 — a genuinely refunded 0% item has `qty` or `refund_total` set, so it passes the guard and its 0% tax is preserved — and it also protects the REST refunds controllers and direct extension callers, which share this exact code path.

## Optional — JS payload hygiene (with a caveat)

We *could* also gate the tax collection like the qty collection, at `meta-boxes-order.js:960`:

```js
$( '.refund input.refund_line_tax' ).each( function( index, item ) {
    if ( $( item ).closest( 'tr' ).data( 'order_item_id' ) && item.value ) {
        // ...
    }
});
```

One caveat worth flagging: the refund-quantity handler writes an explicit `0` into a 0% row's tax field (`meta-boxes-order.js:1125`), so this gate is compatible with the **qty-driven** refund in the repro. But the order-edit and line-total handlers (`416-422`) don't, so a **line-total-driven** refund of a 0% item would leave that field empty and the gate would drop the 0% line — a partial re-break of #27118. So I'd treat this purely as optional payload cleanup and let the **server guard carry correctness**, rather than relying on the JS gate.

## Tests worth adding

- Amount-only refund on a multi-item taxed order → refund has no line items and no tax items.
- Partial refund of one item → no line/tax items created for the siblings.
- Download permissions survive for products that weren't in the refund.
- (Keep the existing 0%-rate test — it's good and still passes with the guard change.)

Happy to pair on this or push the guard change + tests to the branch if that's easier — your call. The core `is_numeric` approach is right; it's just this interaction with the untouched-item guard that needs closing.

**Review by vladolaru — CHANGES_REQUESTED** [review]
Hey Cvetan — lovely bit of archaeology on this one. Pinning a decade-old bug to a falsy-`0` `array_filter` is exactly right, the `is_numeric` instinct is sound, and the comments explaining *why* are genuinely helpful. Two things before this is ready, plus a small optional cleanup.

- The AJAX path has a regression on multi-item / amount-only refunds — untouched items slip past the `wc_create_refund()` guard and become phantom refund line items, which then revoke downloads for products that were never refunded. Full walkthrough + fix snippets in the [dedicated review](https://github.com/woocommerce/woocommerce/pull/66615#pullrequestreview-4703532275).
- The new test only exercises `wc_create_refund()` directly; the AJAX handler that actually runs the #27118 repro has no coverage.
- Optional: the `wc_create_refund()` loop could match the one-line idiom you already used on the AJAX side.

---

<details><summary>Details</summary>

**The regression is the blocker.** The short version: the admin refund JS posts `0` for every item's total and tax (both ungated in `meta-boxes-order.js`), while gating only qty. On trunk the callback-less `array_filter` drops those zeros so untouched items are skipped by the guard at `wc-order-functions.php:620`; with `is_numeric` the `refund_tax` leg stays non-empty, the guard no longer fires, and a negative line item is created for every item in the order. Because those items flow into `$refunded_order_and_products` unconditionally (`wc-order-functions.php:627-632`), a partial refund on an order with downloadables revokes access to products that weren't refunded (`694-704`). It only surfaces on multi-item or amount-only refunds, which is why the single-item repro doesn't catch it. The recommended fix is a server-side guard change (skip an item only when qty, total, and the summed `refund_tax` are all zero) — safe for #27118 and it also covers the REST refunds controllers and direct extension callers. Snippets and the JS-gate caveat are in the linked review.

**Test coverage for the AJAX path.** `test_wc_create_refund_preserves_zero_rate_tax_27118()` calls `wc_create_refund()` with an already-numeric `refund_tax => array( $rate_id => 0 )`, so it verifies that path only. `WC_AJAX::refund_line_items()` — the code that runs when an admin clicks "Refund", i.e. the literal issue repro — has no test around its `array_filter( array_map( 'wc_format_decimal', ... ), 'is_numeric' )` chain. Worth adding coverage there, along with a multi-item / amount-only case and one asserting download permissions survive for unrefunded products.

**Optional idiom consolidation.** The 9-line init/isset/foreach block in `wc_create_refund()` (`wc-order-functions.php:610-618`) is behaviourally identical to `array_map( 'floatval', array_filter( (array) $args['line_items'][ $item_id ]['refund_tax'], 'is_numeric' ) )` — the same concise form you used in the sibling AJAX fix. Purely a consistency nicety; take it or leave it.

**What's genuinely solid.** This closes a real ~9-year-old gap (the falsy-`0` drop traces back to 2016/2017 with no prior fix), the `(float)` cast introduces no precision or stored-type divergence since values re-normalize through `wc_format_refund_total()` before storage, and the `continue` guard reads the raw args so it isn't broken by the change. The core approach is right — it's the untouched-item interaction on the AJAX path that needs closing.

</details>

**Comment by mordeth** [comment]
Hey Vlad, that was a great catch. The single-item repro I verified against had no untouched siblings, so the phantom-item path never fired. I reproduced it exactly as you described before touching anything: pre-fix, an amount-only refund on a two-item order created two phantom negative line items, and a partial refund deleted the download permission for a product that was never refunded.

All three items are addressed:

- **Regression (blocker)**: took your server guard as suggested. The skip condition now checks the sum of the absolute filtered `refund_tax` amounts instead of raw-array emptiness (90dd063bd), so an all-zero tax array can no longer resurrect an item, while a genuine 0% line still survives via its qty or total. Regression tests for all three scenarios you listed (amount-only multi-item, untouched siblings, download survival) failed pre-fix and pass now.
- **AJAX coverage**: added two tests driving `WC_AJAX::refund_line_items()` end to end with the exact payload shape the admin form serializes (2b459f919): a qty-driven refund of a 0% item preserves the tax line (the literal issue repro), and an amount-only multi-item refund creates no phantom items and keeps downloads for unrefunded products. Since the fixes were already in place these couldn't fail-first, so I mutation-verified them instead: reverting the guard fails the amount-only test, and reverting the `is_numeric` filter fails the 0% test.
- **Idiom consolidation**: took it (c6aacd033). Verified it's behaviorally identical first (`array_filter` preserves keys, `floatval` matches the cast).

On the optional JS gate: leaving it out for exactly the caveat you flagged. The line-total handlers don't write an explicit 0 into the tax field, so gating on `item.value` would partially re-break #27118 for total-driven refunds. The server guard carries correctness for the AJAX path, REST controllers, and direct callers alike.

Ready for another look whenever you have a moment.

**Review by vladolaru — APPROVED** [review]
All good now, @mordeth! :shipit:

**Comment by vladolaru** [comment]
Approved — thanks for the thorough follow-up here, Cvetan. Before hitting the button I ran the changes through local mutation testing to independently confirm the fail-on-revert behavior you described, and everything held up cleanly. Sharing the results for the record.

Baseline (all fixes in place): **6/6 pass, 22 assertions.**

| Fix reverted | Failures | Which tests |
|---|---|---|
| **Guard** — `$refund_tax_sum` → raw-array `empty()` (`wc-order-functions.php`) | **4** | The three untouched-item tests + the amount-only AJAX test. Including `test_wc_create_refund_keeps_downloads_of_unrefunded_products` failing with download permissions dropping to `0` — the exact harm the guard prevents. The 0% tests stay green. |
| **AJAX filter** — drop `'is_numeric'` (`class-wc-ajax.php`) | **1** | Only `test_refund_line_items_preserves_zero_rate_tax` (0% line drops, `size 0 vs 1`). |
| **`wc_create_refund()` filter** — drop `'is_numeric'` (`wc-order-functions.php`) | **2** | Both the direct and AJAX 0% tests. |

Each mutation fails *only* the tests asserting the behavior that fix protects, and the suite goes green again with everything restored — so the coverage is tight, not incidental.

One thing the last mutation nicely surfaced: reverting the `wc_create_refund()` filter also fails the *AJAX* 0% test, because the AJAX handler's `$line_items` flow downstream through `wc_create_refund()` and hit its `is_numeric` filter too. So the two filters aren't redundant — they're defense-in-depth, and both are independently covered. Good belt-and-suspenders.

Really nice work closing a decade-old bug without opening a new one. 🚀
